File: code/1456.maximum-number-of-vowels-in-a-substring-of-given-length.php
<?php
/*
 * @lc app=leetcode id=1456 lang=php
 *
 * [1456] Maximum Number of Vowels in a Substring of Given Length
 */

// @lc code=start
class Solution {

    /**
     * @param String $s
     * @param Integer $k
     * @return Integer
     */
    function maxVowels($s, $k) {
        $vowelCount = 0;
        for ($i = 0; $i < $k; $i++) {
            if ($this->isVowel($s[$i])) {
                $vowelCount++;
            }
        }

        $maxCount = $vowelCount;

        for ($i = 1; $i <= strlen($s) - $k; $i++) {
            if ($this->isVowel($s[$i - 1])) {
                $vowelCount--;
            }
            if ($this->isVowel($s[$i + $k - 1])) {
                $vowelCount++;
            }
            if ($vowelCount > $maxCount) {
                $maxCount = $vowelCount;
            }
        }

        return $maxCount;
    }

    function isVowel(string $s): bool
    {
        return str_contains('aeiou', $s);
    }
}
// @lc code=end

File: code/1456.maximum-number-of-vowels-in-a-substring-of-given-length.go
package code

/*
 * @lc app=leetcode id=1456 lang=golang
 *
 * [1456] Maximum Number of Vowels in a Substring of Given Length
 */

// @lc code=start
func maxVowels(s string, k int) int {
	vowels := map[byte]bool{
		'a': true,
		'e': true,
		'i': true,
		'o': true,
		'u': true,
	}

	max := 0
	for i := 0; i < k; i++ {
		if vowels[s[i]] {
			max++
		}
	}

	cur := max
	for i := k; i < len(s); i++ {
		if vowels[s[i-k]] {
			cur--
		}
		if vowels[s[i]] {
			cur++
		}

		if cur > max {
			max = cur
		}
	}

	return max
}

// @lc code=end
